browser/src/control/jsdialog/Util.KeyboardGridNavigation.ts

```diff
@@ -0,0 +1,166 @@
+	const targetElement = focusableElements.find((el) => {
+		const [row, column] = getRowColumn(el);
+		return row === targetRow && column === targetColumn;
+	});
+
+	if (!targetElement && axis === 'vertical') {
+		if (direction === 'next') {
+			// Start from the next sibling of the parent container
+			const nextFocusableElement = findFocusableElement(
+				nextElement as HTMLElement,
+				'next',
+				isFocusable,
+			);
+			if (nextFocusableElement) {
+				nextFocusableElement.focus();
+			}
+		} else if (direction === 'previous') {
+			// Start from the previous sibling of the parent container
+			const previousFocusableElement = findFocusableElement(
+				nextElement as HTMLElement,
+				'previous',
+				isFocusable,
+			);
+			if (previousFocusableElement) {
+				previousFocusableElement.focus();
+			}
+		}
+	}
+
+	if (targetElement) {
+		targetElement.focus();
+	}
+}
+
+function findFocusableElement(
+	element: HTMLElement,
+	direction: 'next' | 'previous',
+	isFocusable: (el: HTMLElement) => boolean,
+): HTMLElement | null {
+	// Check the current element if it is focusable
+	if (isFocusable(element)) return element;
+
+	// Check if sibling is focusable or contains focusable elements
+	const focusableInSibling = findFocusableWithin(
+		element as HTMLElement,
+		direction,
+	);
+	if (focusableInSibling) return focusableInSibling;
+
+	// Depending on the direction, find the next or previous sibling
+	const sibling: Element =
+		direction === 'next'
+			? element.nextElementSibling
+			: element.previousElementSibling;
+
+	if (sibling) {
+		// Recursively check the next or previous sibling of the current sibling
+		return findFocusableElement(sibling as HTMLElement, direction, isFocusable);
+	}
+
+	return null;
+}
+
+// Helper function to find the first focusable element within an element
+function findFocusableWithin(
+	element: HTMLElement,
+	direction: string,
+): HTMLElement | null {
+	const focusableElements = Array.from(element.querySelectorAll('*'));
+	return direction === 'next'
+		? (focusableElements.find(isFocusable) as HTMLElement | null)
+		: (focusableElements.reverse().find(isFocusable) as HTMLElement | null);
+}
+
+// Utility function to check if an element is focusable
+function isFocusable(element: HTMLElement) {
+	if (!element) return false;
+
+	// Check if element is focusable (e.g., input, button, link, etc.)
+	const focusableElements = [
+		'a[href]',
+		'button',
+		'textarea',
+		'input[type="text"]',
+		'input[type="radio"]',
+		'input[type="checkbox"]',
+		'select',
+		'[tabindex]:not([tabindex="-1"])',
+	];
+
+	return focusableElements.some((selector) => element.matches(selector));
+}
+
+function getRowColumn(element: HTMLElement): [number, number] {
+	const index = element.getAttribute('index');
+	if (!index) return [-1, -1]; // we will never have this kind of index this is why we are pssing nagative values here
+	const [row, column] = index.split(':').map(Number);
+	return [row, column];
+}
```

Comment:
at the end please add "exported" / public functions:
`JSDialog.nameOfFuntion = localNameOfFunction;`
and in the other files please use that `JSDialog.nameOfFunction` to call it.

I can imagine we should need only one function to be exported, so we can add this functionality to the widget just by one line:
`JSDialog.KeyboardGridNavigation(<some grid container>, <some additional args>);`

Reply:
> I can imagine we should need only one function to be exported, so we can add this functionality to the widget just by one line:
> `JSDialog.KeyboardGridNavigation(<some grid container>, <some additional args>);`

I am doing R and D on this suggestion

Reply:
Done 🎉